### DISCUSSION TITLE
TypeScript support for Near Social VM components

### DISCUSSION THREAD

**Discussion by mpeterdev** [NearSocial/viewer#110] [General]
Are we interested in adding TypeScript support for component authoring? It is possible to transform TypeScript in the browser to JS in a way that is compatible with the VM via a config option to preserve JSX (see [here](https://www.typescriptlang.org/docs/handbook/jsx.html))

Performance wasn't ideal in my POC, but that could likely be worked out.

Concerns:
We would probably need to save two sets of source to SocialDB: the TypeScript and the transpiled JS.

- We need to store JS because I think it is unlikely we could transpile on the fly when viewing a component for performance reasons. 
- We need to store the TS because that is the source that will be needed when editing the component.

Storing both would about double the space required on-chain for a component.

Possible attack vector: Component author saves JS which does not match the TS, adding in malicious code. A user audits only the TS and believes the component to be safe

**Comment by evgenykuzyakov** [comment]
Can we convert TS on the fly and then cache it to indexedDB?

**Reply by petarvujovic98** [reply]
I think that `tsc` could be used for this, although it would be heavy to include. It does however allow you to compile TypeScript in the browser. Another solution would be to run the compilation on the server and send back the compiled JS, but this does introduce some potential vulnerabilities. The TS playground is an example of this
